File: typescript/allergies/allergies.ts
const allergens: { [key: string]: number } = {
  eggs: 1,
  peanuts: 2,
  shellfish: 4,
  strawberries: 8,
  tomatoes: 16,
  chocolate: 32,
  pollen: 64,
  cats: 128
}

export class Allergies {
  allergenIndex: number

  constructor(allergenIndex: number) {
    this.allergenIndex = allergenIndex
  }

  public list(): string[] {
    return Object.keys(allergens).filter((allergen: string) => this.allergicTo(allergen))
  }

  public allergicTo(allergen: string): boolean {
    return Boolean(this.allergenIndex & allergens[allergen])
  }
}

File: ruby/allergies/allergies.rb
require 'pry'

class Allergies
  def initialize(score)
    @allergen_hash = create_allergen_hash(score)
  end

  def allergic_to?(allergen)
    @allergen_hash[allergen] == '1'
  end

  def list
    @list ||= @allergen_hash.filter do |allergen, positive|
      positive == '1'
    end.keys
  end

  private

  def create_allergen_hash(score)
    digits = score.to_s(2).reverse.split('')[0..9]
    allergens = [
      'eggs',
      'peanuts', 
      'shellfish',
      'strawberries',
      'tomatoes',
      'chocolate',
      'pollen',
      'cats',
    ]
    allergens.zip(digits).to_h
  end
end
